tests: Port `translation` to rmake.rs

jieyouxu opened the pull request:
Part of #121876.

This PR partially supersedes #129011 and is co-authored with @Oneirical.

## Summary

This PR ports `tests/run-make/translation` to rmake.rs. Notable changes from the Makefile version include:

- We now actually fail if the rustc invocations fail... The Makefile did not have `SHELL=/bin/bash -o pipefail`, so all the piped rustc invocations to grep vacuously succeeded, even if the broken ftl test case actually regressed over time and ICEs on current master.
    - That test case is converted to assert it fails with a FIXME backlinking to #135817.
- The test coverage is expanded to not ignore windows. Instead, the test now uses symlink capability detection to gate test execution.
- Added some backlinks to relevant tracking issues and the initial translation infra implementation PR.

## Review advice

Best reviewed commit-by-commit.

r? compiler

try-job: aarch64-apple
try-job: i686-mingw

jieyouxu committed:
run-make-support: add `sysroot` helper

Convenience helper for `rustc --print=sysroot`.

jieyouxu committed:
tests: port `translation` to rmake.rs

Co-authored-by: Oneirical <manchot@videotron.ca>

jieyouxu commented:
r? translation

jieyouxu commented:
Based on triagebot... maybe r? @davidtwco

bors commented:
:pushpin: Commit 139c5ae0e943f0151f41bd8b010b8a51866f1482 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 139c5ae0e943f0151f41bd8b010b8a51866f1482 -->
<!-- homu: {"type":"Approved","sha":"139c5ae0e943f0151f41bd8b010b8a51866f1482","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jieyouxu commented:
Failed in https://github.com/rust-lang/rust/pull/135861#issuecomment-2606505618
@bors r-

jieyouxu committed:
run-make-support: improve docs for `assert_exit_code`

jieyouxu commented:
Slightly documented `assert_exit_code` and changed `assert` -> `assert_eq` to have a better panic message, added `//@ needs-process` to the test because `wasm32-wasip1` is a great target, but no functional changes otherwise.

@bors r=compiler-errors rollup=iffy

bors commented:
:pushpin: Commit e142d0c38049b873926ed337fe3697713f98eb73 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors e142d0c38049b873926ed337fe3697713f98eb73 -->
<!-- homu: {"type":"Approved","sha":"e142d0c38049b873926ed337fe3697713f98eb73","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jieyouxu commented:
@bors r-

jieyouxu commented:
Apparently `//@ needs-subprocess` isn't a thing yet cf. #128295, oops.

@bors r=compiler-errors

bors commented:
:pushpin: Commit 6510e9914ecdc4e786a1faac2f95c58366af4098 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 6510e9914ecdc4e786a1faac2f95c58366af4098 -->
<!-- homu: {"type":"Approved","sha":"6510e9914ecdc4e786a1faac2f95c58366af4098","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

jieyouxu commented on tests/run-make/translation/rmake.rs:
```diff
@@ -0,0 +1,178 @@
+// This test uses symbolic links to stub out a fake sysroot to save testing time.
+//@ needs-symlink
+// Needs to spawn `rustc` subprocesses.
+// FIXME(#128295): clean these ignores up in favor of `needs-subprocess` directive
+//@ ignore-wasm
+//@ ignore-emscripten
+//@ ignore-sgx
```
I really should implement this lol

jieyouxu replied:
#135926, I'll send a follow-up after *that* merges, didn't want to block this PR on *that* PR

jieyouxu replied:
Fixed

jieyouxu commented:
@bors r- (might as well fix that since this hasn't reached bors yet)

jieyouxu commented:
Uses `//@ needs-subprocess` now.

@bors r=compiler-errors

bors commented:
:pushpin: Commit 388a52f55151186731eec45f52a3786112ee3dcc has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 388a52f55151186731eec45f52a3786112ee3dcc -->
<!-- homu: {"type":"Approved","sha":"388a52f55151186731eec45f52a3786112ee3dcc","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r-
https://github.com/rust-lang/rust/pull/136051#issuecomment-2613963923 looks like its still failing

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit 4e3860654c642e5ad9998d8c9670a74e12734b3b with merge 878483661bc50b0b3540145d28675a78cd75c10d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"4e3860654c642e5ad9998d8c9670a74e12734b3b","merge_sha":"878483661bc50b0b3540145d28675a78cd75c10d"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12965715365/job/36165865907)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/12965715365/job/36165865907","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`aarch64-apple`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12965715365/job/36165752458) [(plain)](https://github.com/rust-lang-ci/rust/commit/878483661bc50b0b3540145d28675a78cd75c10d/checks/36165752458/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  98% (50608/51640)
Updating files:  99% (51124/51640)
Updating files: 100% (51640/51640)
Updating files: 100% (51640/51640), done.
Switched to a new branch 'try'
branch 'try' set up to track 'origin/try'.
[command]/opt/homebrew/bin/git log -1 --format=%H
878483661bc50b0b3540145d28675a78cd75c10d
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
fmt check
fmt: skipped untracked file AWSCLIV2.pkg
fmt: checked 5808 files
tidy check
tidy error: /Users/runner/work/rust/rust/tests/run-make/translation/rmake.rs:66: Instead of XXX use FIXME
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Sat Jan 25 14:27:54 UTC 2025
  network time: Sat, 25 Jan 2025 14:27:54 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit 67357fee1d6645af303899c25a54998a6479d2c7 with merge 93fdfc404e0e6edf2ae4f08380cdc5293b1c914e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"67357fee1d6645af303899c25a54998a6479d2c7","merge_sha":"93fdfc404e0e6edf2ae4f08380cdc5293b1c914e"} -->

rust-log-analyzer commented:
The job **`aarch64-apple`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12965826776/job/36165963048) [(plain)](https://github.com/rust-lang-ci/rust/commit/93fdfc404e0e6edf2ae4f08380cdc5293b1c914e/checks/36165963048/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  98% (50609/51641)
Updating files:  99% (51125/51641)
Updating files: 100% (51641/51641)
Updating files: 100% (51641/51641), done.
Switched to a new branch 'try'
branch 'try' set up to track 'origin/try'.
[command]/opt/homebrew/bin/git log -1 --format=%H
93fdfc404e0e6edf2ae4f08380cdc5293b1c914e
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
---- [run-make] tests/run-make/translation stdout ----

error: rmake recipe failed to complete
status: exit status: 101
command: cd "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/run-make/translation/rmake_out" && env -u RUSTFLAGS AR="ar" CARGO="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage1-tools-bin/cargo" CC="/Applications/Xcode_15.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx11 -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk" CXX="/Applications/Xcode_15.4.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx11 -isysroot /Applications/Xcode_15.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk -stdlib=libc++" DYLD_LIBRARY_PATH="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0-bootstrap-tools/aarch64-apple-darwin/release/deps:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0/lib:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2-tools-bin:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib/rustlib/aarch64-apple-darwin/lib" HOST_RPATH_DIR="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib" LD_LIB_PATH_ENVVAR="DYLD_LIBRARY_PATH" LLVM_BIN_DIR="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/ci-llvm/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegendata codegentypes core coroutines coverage csky cskyasmparser cskycodegen cskydesc cskydisassembler cskyinfo debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld sandboxir scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/ci-llvm/bin/FileCheck" NODE="/opt/homebrew/bin/node" PYTHON="/usr/bin/python3" RUSTC="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/bin/rustc" RUSTDOC="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/bin/rustdoc" SOURCE_ROOT="/Users/runner/work/rust/rust" TARGET="aarch64-apple-darwin" TARGET_RPATH_DIR="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage2/lib/rustlib/aarch64-apple-darwin/lib" TARGET_RPATH_ENV="/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/run-make/translation/rmake_out:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0-bootstrap-tools/aarch64-apple-darwin/release/deps:/Users/runner/work/rust/rust/build/aarch64-apple-darwin/stage0/lib" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/run-make/translation/rmake"
--- stderr -------------------------------
stdout: ""
stdout: ""
stderr: "error: struct literal body without path\n  --> test.rs:8:17\n   |\n8  |   fn foo() -> Foo {\n   |  _________________^\n9  | |     val: (),\n10 | | }\n   | |_^\n   |\nhelp: this is a test suggestion\n   |\n8  ~ fn foo() -> Foo { SomeStruct {\n9  |     val: (),\n10 ~ } }\n   |\n\nerror: struct literal body without path\n  --> test.rs:15:13\n   |\n15 |       let x = {\n   |  _____________^\n16 | |         val: (),\n17 | |     };\n   | |_____^\n   |\nhelp: this is a test suggestion\n   |\n15 ~     let x = { SomeStruct {\n16 |         val: (),\n17 ~     } };\n   |\n\nerror[E0308]: mismatched types\n  --> test.rs:14:14\n   |\n14 |     x.val == 42;\n   |     -----    ^^ expected `()`, found integer\n   |     |\n   |     expected because this is `()`\n\nerror: aborting due to 3 previous errors\n\nFor more information about this error, try `rustc --explain E0308`.\n"
thread 'main' panicked at /Users/runner/work/rust/rust/tests/run-make/translation/rmake.rs:69:13:
assertion `left == right` failed
  left: Some(1)
 right: Some(101)
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12965826776/job/36166939485)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/12965826776/job/36166939485","builder_name":"checks-actions"} -->

jieyouxu commented on tests/run-make/translation/rmake.rs:
```diff
@@ -0,0 +1,173 @@
+    // `broken.ftl`, even though the original intention seems to be that it should gracefully
+    // failover to the fallback bundle. On `aarch64-apple-darwin`, somehow it *doesn't* ICE.
+
+    rustc()
+        .env("RUSTC_ICE", "0") // disable ICE dump file, not needed
+        .input("test.rs")
+        .arg("-Ztranslate-additional-ftl=broken.ftl")
+        .run_fail();
```
I'm not going to assert that this ICEs because it somehow *doesn't* ICE on `aarch64-apple-darwin`, instead I'm just going to assert that it fails.

jieyouxu commented:
@bors r=compiler-errors rollup=never

bors commented:
:pushpin: Commit 33851665a421d81cc5ff275edb5cc8f117e12007 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 33851665a421d81cc5ff275edb5cc8f117e12007 -->
<!-- homu: {"type":"Approved","sha":"33851665a421d81cc5ff275edb5cc8f117e12007","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 33851665a421d81cc5ff275edb5cc8f117e12007 with merge 288b058e1dfa7f111f726e8b42c48b872c5f19a5...
<!-- homu: {"type":"BuildStarted","head_sha":"33851665a421d81cc5ff275edb5cc8f117e12007","merge_sha":"288b058e1dfa7f111f726e8b42c48b872c5f19a5"} -->

rust-log-analyzer commented:
The job **`i686-mingw`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/12977117608/job/36190190544) [(plain)](https://github.com/rust-lang-ci/rust/commit/288b058e1dfa7f111f726e8b42c48b872c5f19a5/checks/36190190544/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [run-make] tests\run-make\translation stdout ----

error: rmake recipe failed to complete
status: exit code: 101
command: "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\test\\run-make\\translation\\rmake.exe"
--- stderr -------------------------------

thread 'main' panicked at C:\a\rust\rust\tests\run-make\translation\rmake.rs:123:24:
thread 'main' panicked at C:\a\rust\rust\tests\run-make\translation\rmake.rs:123:24:
failed to remove file at `C:\a\rust\rust\build\i686-pc-windows-gnu\test\run-make\translation\temporary-directory\fakeroot\lib`: Access is denied. (os error 5)
------------------------------------------




failures:
    [run-make] tests\run-make\translation

test result: FAILED. 275 passed; 1 failed; 102 ignored; 0 measured; 6 filtered out; finished in 128.37s

Some tests failed in compiletest suite=run-make mode=run-make host=i686-pc-windows-gnu target=i686-pc-windows-gnu
Build completed unsuccessfully in 1:40:20
make: *** [Makefile:120: ci-mingw-x] Error 1
  network time: Sun, 26 Jan 2025 19:56:17 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12977117608/job/36191735432)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/12977117608/job/36191735432","builder_name":"checks-actions"} -->

jieyouxu commented:
@rustbot author

jieyouxu commented:
I love Windows and symlinks.
@bors r=compiler-errors

bors commented:
:pushpin: Commit 717b16ec0e8a7335976309cb77a313c3f441e090 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 717b16ec0e8a7335976309cb77a313c3f441e090 -->
<!-- homu: {"type":"Approved","sha":"717b16ec0e8a7335976309cb77a313c3f441e090","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bjorn3 commented on src/tools/run-make-support/src/command.rs:
```diff
@@ -388,9 +388,15 @@ impl CompletedProcess {
     pub fn assert_exit_code(&self, code: i32) -> &Self {
-        assert!(self.output.status.code() == Some(code));
+        // FIXME(jieyouxu): this should really be named `exit_status`, because std has an `ExitCode`
+        // that means a different thing.
```
`ExitStatus` is effectively an enum between an exit code and information about the signal that caused the process to exit. This function checks that the `ExitStatus` is an exit code and in particular that it is the given exit code.

jieyouxu replied:
Yeah, this FIXME was outdated from debugging, I'll send a follow-up to remove it (the asserion behaves as intended)

jieyouxu replied:
Nvm, fixed

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13010398278/job/36293051906)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13010398278/job/36293051906","builder_name":"checks-actions"} -->

jieyouxu commented:
@bors r-

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit 6f5a1035f8b3694e519598ddb27d884a591e0d32 with merge a459f7545933b917f836764607766733d0060a9d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"6f5a1035f8b3694e519598ddb27d884a591e0d32","merge_sha":"a459f7545933b917f836764607766733d0060a9d"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13027096936/job/36344713308)
Build commit: a459f7545933b917f836764607766733d0060a9d (`a459f7545933b917f836764607766733d0060a9d`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13027096936/job/36344713308"},"merge_sha":"a459f7545933b917f836764607766733d0060a9d"} -->

jieyouxu commented:
@bors r=compiler-errors

bors commented:
:pushpin: Commit 6f5a1035f8b3694e519598ddb27d884a591e0d32 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 6f5a1035f8b3694e519598ddb27d884a591e0d32 -->
<!-- homu: {"type":"Approved","sha":"6f5a1035f8b3694e519598ddb27d884a591e0d32","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 6f5a1035f8b3694e519598ddb27d884a591e0d32 with merge 4a5f1cc52b7546c61462f1d92b60cc80d40e170f...
<!-- homu: {"type":"BuildStarted","head_sha":"6f5a1035f8b3694e519598ddb27d884a591e0d32","merge_sha":"4a5f1cc52b7546c61462f1d92b60cc80d40e170f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13041256234/job/36389521905)
Approved by: compiler-errors
Pushing 4a5f1cc52b7546c61462f1d92b60cc80d40e170f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13041256234/job/36389521905"},"merge_sha":"4a5f1cc52b7546c61462f1d92b60cc80d40e170f"} -->

rust-timer commented:
Finished benchmarking commit (4a5f1cc52b7546c61462f1d92b60cc80d40e170f): [comparison URL](https://perf.rust-lang.org/compare.html?start=ae5de6c759cd337ecdb2de4e94f47eaafb5d4606&end=4a5f1cc52b7546c61462f1d92b60cc80d40e170f&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ae5de6c759cd337ecdb2de4e94f47eaafb5d4606&end=4a5f1cc52b7546c61462f1d92b60cc80d40e170f&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.3% | [0.3%, 0.3%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ae5de6c759cd337ecdb2de4e94f47eaafb5d4606&end=4a5f1cc52b7546c61462f1d92b60cc80d40e170f&stat=max-rss)
<details>
<summary>Results (primary -2.3%, secondary -2.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.3% | [-2.4%, -2.2%] | 3     |
| Improvements ✅ <br /> (secondary) | -2.8% | [-2.8%, -2.8%] | 1     |
| All ❌✅ (primary)                 | -2.3% | [-2.4%, -2.2%] | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ae5de6c759cd337ecdb2de4e94f47eaafb5d4606&end=4a5f1cc52b7546c61462f1d92b60cc80d40e170f&stat=cycles:u)
<details>
<summary>Results (secondary 2.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.2% | [2.2%, 2.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=ae5de6c759cd337ecdb2de4e94f47eaafb5d4606&end=4a5f1cc52b7546c61462f1d92b60cc80d40e170f&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 777.289s -> 775.772s (-0.20%)
**Artifact size**: 328.44 MiB -> 328.49 MiB (0.02%)
